File: web/admin/app/store/book.ts
import { Action, Module, Mutation, VuexModule } from 'vuex-module-decorators'
import axios, { AxiosResponse } from 'axios'
import { IBook, IBookState } from '~/types/store'
import { IErrorResponse, IRakutenBookItem, IRakutenBookResponse } from '~/types/responses'
import { ApiError } from '~/types/exception'
import { IBookSearchForm } from '~/types/forms'

const initialState: IBookState = {
  books: [],
}

@Module({
  name: 'book',
  stateFactory: true,
  namespaced: true,
})
export default class BookModule extends VuexModule {
  private books: IBook[] = initialState.books

  public get getBooks(): IBook[] {
    return this.books
  }

  @Mutation
  private setBooks(books: IBook[]): void {
    this.books = books
  }

  @Action({ rawError: true })
  public searchBookFromRakutenBooksAPI(payload: IBookSearchForm): Promise<void> {
    const url: string = getRakutenBooksURL(payload)

    return new Promise((resolve: () => void, reject: (reason: Error) => void) => {
      axios
        .get(url)
        .then((res: AxiosResponse<IRakutenBookResponse>) => {
          const data = res.data
          if (!data) {
            return
          }

          const books: IBook[] = data.Items.map((item: IRakutenBookItem) => {
            return {
              id: 0,
              title: item.title,
              titleKana: item.titleKana,
              description: item.itemCaption,
              isbn: item.isbn,
              publisher: item.publisherName,
              publishedOn: item.salesDate,
              thumbnailUrl: item.largeImageUrl,
              rakutenUrl: item.itemUrl,
              rakutenGenreId: item.booksGenreId,
              author: item.author,
              authorKana: item.authorKana,
              createdAt: '',
              updatedAt: '',
            }
          })

          this.setBooks(books)
          resolve()
        })
        .catch((err: any) => {
          const { data, status }: IErrorResponse = err.response
          reject(new ApiError(status, data.message, data))
        })
    })
  }
}

function getRakutenBooksURL(form: IBookSearchForm): string {
  const applicationId: string = process.env.rakutenAppId || ''
  let baseURL: string = `https://app.rakuten.co.jp/services/api/BooksBook/Search/20170404?format=json&formatVersion=2&applicationId=${applicationId}`

  if (!form?.params) {
    return baseURL
  }

  if (form.params.title !== '') {
    baseURL = `${baseURL}&title=${form.params.title}`
  }

  if (form.params.author !== '') {
    baseURL = `${baseURL}&author=${form.params.author}`
  }

  if (form.params.publisher !== '') {
    baseURL = `${baseURL}&publisherName=${form.params.publisher}`
  }

  if (form.params.isbn !== '') {
    baseURL = `${baseURL}&isbn=${form.params.isbn}`
  }

  if (form.params.size !== 0) {
    baseURL = `${baseURL}&size=${form.params.size}`
  }

  return baseURL
}

File: api/internal/gateway/native/v1/entity/book.go
package entity

import (
	"strings"

	"github.com/calmato/gran-book/api/internal/gateway/entity"
)

type Book struct {
	ID           int64      `json:"id"`           // 書籍ID
	Title        string     `json:"title"`        // タイトル
	TitleKana    string     `json:"titleKana"`    // タイトル(かな)
	Description  string     `json:"description"`  // 説明
	Isbn         string     `json:"isbn"`         // ISBN
	Publisher    string     `json:"publisher"`    // 出版社名
	PublishedOn  string     `json:"published_on"` // 出版日
	ThumbnailURL string     `json:"thumbnailUrl"` // サムネイルURL
	RakutenURL   string     `json:"rakutenUrl"`   // 楽天ショップURL
	Size         string     `json:"size"`         // 楽天書籍サイズ
	Author       string     `json:"author"`       // 著者名一覧
	AuthorKana   string     `json:"authorKana"`   // 著者名一覧(かな)
	Bookshelf    *Bookshelf `json:"bookshelf"`    // ユーザーの本棚情報
	CreatedAt    string     `json:"createdAt"`    // 登録日時
	UpdatedAt    string     `json:"updatedAt"`    // 更新日時
}

type Books []*Book

func NewBook(b *entity.Book, bs *entity.Bookshelf) *Book {
	return &Book{
		ID:           b.Id,
		Title:        b.Title,
		TitleKana:    b.TitleKana,
		Description:  b.Description,
		Isbn:         b.Isbn,
		Publisher:    b.Publisher,
		PublishedOn:  b.PublishedOn,
		ThumbnailURL: b.ThumbnailUrl,
		RakutenURL:   b.RakutenUrl,
		Size:         b.RakutenSize,
		Author:       strings.Join(b.AuthorNames(), "/"),
		AuthorKana:   strings.Join(b.AuthorNameKanas(), "/"),
		CreatedAt:    b.CreatedAt,
		UpdatedAt:    b.UpdatedAt,
		Bookshelf:    NewBookshelf(bs),
	}
}

func NewBooks(bm map[int64]*entity.Book, bss entity.Bookshelves) Books {
	res := make(Books, 0, len(bss))
	for _, bs := range bss {
		b, ok := bm[bs.BookId]
		if !ok {
			continue
		}
	}
}
